File: src/runtime/control.ts
import type { ApolloRequestRecord } from '../core/types.js';

type FetchOptions = { method?: string; body?: unknown; headers?: Record<string, string>; auth?: 'session' | 'bearer' | 'none' };

function segment(value: string): string { return encodeURIComponent(value); }

export class ApolloControlClient {
  private cookie = '';
  constructor(
    public readonly portalUrl: string,
    public readonly configServiceUrl: string,
    private readonly observer?: ApolloRequestRecord[],
  ) {}

  async login(username = 'apollo', password = 'admin'): Promise<void> {
    const response = await fetch(`${this.portalUrl}/signin`, {
      method: 'POST', redirect: 'manual',
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({ username, password }),
    });
    const setCookie = response.headers.get('set-cookie');
    if (!setCookie || (response.status !== 302 && response.status !== 200)) throw new Error(`Portal login failed: ${response.status}`);
    this.cookie = setCookie.split(';', 1)[0]!;
  }

  clearSession(): void { this.cookie = ''; }

  async request<T = unknown>(path: string, options: FetchOptions = {}): Promise<T> {
    const headers: Record<string, string> = { accept: 'application/json', ...options.headers };
    if (options.body !== undefined) headers['content-type'] = 'application/json';
    if ((options.auth ?? 'session') === 'session') headers.cookie = this.cookie;
    const started = new Date().toISOString();
    const response = await fetch(`${this.portalUrl}${path}`, {
      method: options.method ?? 'GET', headers,
      body: options.body === undefined ? undefined : JSON.stringify(options.body),
    });
    this.observer?.push({ timestamp: started, method: options.method ?? 'GET', path, status: response.status, authType: headers.cookie ? 'cookie' : headers.authorization ? 'bearer' : 'none' });
    const text = await response.text();
    if (!response.ok) throw new Error(`${options.method ?? 'GET'} ${path}: ${response.status} ${text.slice(0, 500)}`);
    if (!text) return undefined as T;
    try { return JSON.parse(text) as T; } catch { return text as T; }
  }

  async createApp(appId: string, name = appId): Promise<void> {
    await this.request('/apps', { method: 'POST', body: { appId, name, orgId: 'TEST1', orgName: '样例部门1', ownerName: 'apollo', admins: ['apollo'] } });
  }

  async createUserToken(name: string, appIds: string[]): Promise<string> {
    const result = await this.request<{ tokenValue: string }>('/openapi/v1/user-tokens', { method: 'POST', body: { name, appIds, envs: ['LOCAL'] } });
    return result.tokenValue;
  }

  namespacePath(appId: string, namespace = 'application'): string {
    return `/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters/default/namespaces/${segment(namespace)}`;
  }

  async createAppNamespace(appId: string, namespace: string, format: string): Promise<void> {
    await this.request(`/openapi/v1/apps/${segment(appId)}/appnamespaces`, { method: 'POST', body: { name: namespace, appId, format, isPublic: false, appendNamespacePrefix: true, comment: 'apollo-evals' } });
  }

  async createNamespace(appId: string, namespace: string): Promise<void> {
    try { await this.request('/openapi/v1/namespaces', { method: 'POST', body: [{ appId, env: 'LOCAL', clusterName: 'default', appNamespaceName: namespace }] }); }
    catch (error) {
      if (!String(error).includes('create namespace failed for')) throw error;
      await this.request(`${this.namespacePath(appId, namespace)}?fillItemDetail=false`);
    }
  }

  async createCluster(appId: string, clusterName: string): Promise<void> {
    await this.request(`/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters?operator=apollo`, { method: 'POST', body: { appId, name: clusterName, dataChangeCreatedBy: 'apollo', dataChangeLastModifiedBy: 'apollo' } });
  }

  async createNamespaceInCluster(appId: string, clusterName: string, namespace: string): Promise<void> {
    try { await this.request('/openapi/v1/namespaces', { method: 'POST', body: [{ appId, env: 'LOCAL', clusterName, appNamespaceName: namespace }] }); }
    catch (error) {
      if (!String(error).includes('create namespace failed for')) throw error;
      await this.request(`/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters/${segment(clusterName)}/namespaces/${segment(namespace)}?fillItemDetail=false`);
    }
  }

  async putItem(appId: string, namespace: string, key: string, value: string, type = 0): Promise<void> {
    const path = `${this.namespacePath(appId, namespace)}/items/${segment(key)}?createIfNotExists=true&operator=apollo`;
    const body = { key, value, type, dataChangeCreatedBy: 'apollo', dataChangeLastModifiedBy: 'apollo' };
    try { await this.request(path, { method: 'PUT', body }); }
    catch (error) {
      if (!/\b404\b/.test(String(error))) throw error;
      await this.request(`${this.namespacePath(appId, namespace)}/items?operator=apollo`, { method: 'POST', body });
    }
  }

  async putItemInCluster(appId: string, clusterName: string, namespace: string, key: string, value: string, type = 0): Promise<void> {
    const base = `/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters/${segment(clusterName)}/namespaces/${segment(namespace)}`;
    const body = { key, value, type, dataChangeCreatedBy: 'apollo', dataChangeLastModifiedBy: 'apollo' };
    try { await this.request(`${base}/items/${segment(key)}?createIfNotExists=true&operator=apollo`, { method: 'PUT', body }); }
    catch (error) {
      if (!/\b404\b/.test(String(error))) throw error;
      await this.request(`${base}/items?operator=apollo`, { method: 'POST', body });
    }
  }

  async deleteItem(appId: string, namespace: string, key: string): Promise<void> {
    await this.request(`${this.namespacePath(appId, namespace)}/items/${segment(key)}?operator=apollo`, { method: 'DELETE' });
  }

  async putText(appId: string, namespace: string, configText: string): Promise<void> {
    const extension = namespace.includes('.') ? namespace.slice(namespace.lastIndexOf('.') + 1) : 'properties';
    await this.request(`${this.namespacePath(appId, namespace)}/items?operator=apollo`, { method: 'PUT', body: { appId, env: 'LOCAL', clusterName: 'default', namespaceName: namespace, format: extension, configText, operator: 'apollo' } });
  }

  async release(appId: string, namespace: string, title: string): Promise<Record<string, unknown>> {
    return await this.request(`${this.namespacePath(appId, namespace)}/releases?operator=apollo`, { method: 'POST', body: { releaseTitle: title, releaseComment: 'apollo-evals', releasedBy: 'apollo', isEmergencyPublish: false } });
  }

  async rollback(releaseId: number, toReleaseId?: number): Promise<void> {
    const suffix = toReleaseId === undefined ? '' : `&toReleaseId=${toReleaseId}`;
    await this.request(`/openapi/v1/envs/LOCAL/releases/${releaseId}/rollback?operator=apollo${suffix}`, { method: 'PUT' });
  }

  async items(appId: string, namespace = 'application'): Promise<Array<{ key: string; value: string; type: number }>> {
    try {
      const page = await this.request<{ content: Array<{ key: string; value: string; type: number }> }>(`${this.namespacePath(appId, namespace)}/items?page=0&size=500`);
      return page.content ?? [];
    } catch (error) { if (String(error).includes('404')) return []; throw error; }
  }

  async latestRelease(appId: string, namespace = 'application'): Promise<Record<string, unknown> | null> {
    try { return await this.request<Record<string, unknown>>(`${this.namespacePath(appId, namespace)}/releases/latest`); } catch (error) {
      if (String(error).includes('404')) return null;
      throw error;
    }
  }

  async activeReleases(appId: string, namespace = 'application'): Promise<Array<Record<string, unknown>>> {
    const page = await this.request<{ content: Array<Record<string, unknown>> }>(`${this.namespacePath(appId, namespace)}/releases/active?page=0&size=100`);
    return page.content ?? [];
  }

  async itemsInCluster(appId: string, clusterName: string, namespace = 'application'): Promise<Array<{ key: string; value: string; type: number }>> {
    const page = await this.request<{ content: Array<{ key: string; value: string; type: number }> }>(`/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters/${segment(clusterName)}/namespaces/${segment(namespace)}/items?page=0&size=500`);
    return page.content ?? [];
  }

  async releaseInCluster(appId: string, clusterName: string, namespace: string, title: string): Promise<Record<string, unknown>> {
    return await this.request(`/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters/${segment(clusterName)}/namespaces/${segment(namespace)}/releases?operator=apollo`, { method: 'POST', body: { releaseTitle: title, releaseComment: 'apollo-evals', releasedBy: 'apollo', isEmergencyPublish: false } });
  }

  async latestReleaseInCluster(appId: string, clusterName: string, namespace = 'application'): Promise<Record<string, unknown> | null> {
    try { return await this.request(`/openapi/v1/envs/LOCAL/apps/${segment(appId)}/clusters/${segment(clusterName)}/namespaces/${segment(namespace)}/releases/latest`); } catch (error) {
      if (String(error).includes('404')) return null;
      throw error;
    }
  }

  async appNamespaces(appId: string): Promise<Array<Record<string, unknown>>> {
    return await this.request(`/openapi/v1/apps/${segment(appId)}/appnamespaces`);
  }

  async config(appId: string, namespace = 'application'): Promise<Record<string, string>> {
    const response = await fetch(`${this.configServiceUrl}/configs/${segment(appId)}/default/${segment(namespace)}`);
    if (response.status === 404) return {};
    if (!response.ok) throw new Error(`Config Service read failed: ${response.status}`);
    const body = await response.json() as { configurations: Record<string, string> };
    return body.configurations;
  }
}

File: apollo_testkit/control.py
"""Administrative API used only inside the fixture and verifier services."""

import json
import time
import urllib.error
import urllib.parse
import urllib.request


class APIError(RuntimeError):
    def __init__(self, status, message):
        self.status = status
        super().__init__(f"Apollo HTTP {status}: {message[:400]}")


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


class Apollo:
    def __init__(self, host="apollo"):
        self.portal = f"http://{host}:8070"
        self.config_service = f"http://{host}:8080"
        self.cookie = ""

    def login(self):
        request = urllib.request.Request(
            self.portal + "/signin",
            data=b"username=apollo&password=admin",
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )
        try:
            response = urllib.request.build_opener(NoRedirect).open(request, timeout=15)
        except urllib.error.HTTPError as error:
            response = error
        cookie = response.headers.get("Set-Cookie", "")
        if response.code not in (200, 302) or not cookie:
            raise APIError(response.code, "Portal login failed")
        self.cookie = cookie.split(";", 1)[0]
        response.close()

    def request(self, path, method="GET", body=None, *, config=False):
        headers = {"Accept": "application/json"}
        if not config:
            headers["Cookie"] = self.cookie
        data = None
        if body is not None:
            headers["Content-Type"] = "application/json"
            data = json.dumps(body, separators=(",", ":")).encode()
        request = urllib.request.Request(
            (self.config_service if config else self.portal) + path,
            data=data,
            headers=headers,
            method=method,
        )
        try:
            with urllib.request.urlopen(request, timeout=30) as response:
                raw = response.read().decode()
        except urllib.error.HTTPError as error:
            raise APIError(error.code, error.read().decode(errors="replace")) from error
        return json.loads(raw) if raw else None

    def ready(self, timeout=180):
        deadline = time.monotonic() + timeout
        last = None
        while time.monotonic() < deadline:
            try:
                self.request("/health", config=True)
                self.login()
                try:
                    self.request(self.ns("SampleApp"))
                except APIError as error:
                    if error.status != 404:
                        raise
                return
            except (OSError, APIError, ValueError) as error:
                last = error
                time.sleep(1)
        raise TimeoutError(f"Apollo did not become ready: {last}")

    @staticmethod
    def ns(app, namespace="application", cluster="default"):
        q = lambda value: urllib.parse.quote(str(value), safe="")
        return f"/openapi/v1/envs/LOCAL/apps/{q(app)}/clusters/{q(cluster)}/namespaces/{q(namespace)}"

    def create_app(self, app):
        self.request(
            "/apps",
            "POST",
            {
                "appId": app,
                "name": app,
                "orgId": "TEST1",
                "orgName": "样例部门1",
                "ownerName": "apollo",
                "admins": ["apollo"],
            },
        )

    def token(self, app, *, scoped=False):
        body = {"name": f"eval-{app}", "appIds": [app], "envs": ["LOCAL"]}
        if scoped:
            body.update(
                operations=["config:read", "config:modify", "config:release"],
                namespaces=[
                    {
                        "appId": app,
                        "env": "LOCAL",
                        "clusterName": "default",
                        "namespaceName": "application",
                    }
                ],
            )
        return self.request("/openapi/v1/user-tokens", "POST", body)["tokenValue"]

    def app_namespace(self, app, name, fmt="properties", public=False):
        self.request(
            f"/openapi/v1/apps/{app}/appnamespaces",
            "POST",
            {
                "name": name,
                "appId": app,
                "format": fmt,
                "isPublic": public,
                "appendNamespacePrefix": not public,
                "comment": "apollo-evals",
            },
        )

    def namespace(self, app, namespace, cluster="default"):
        try:
            self.request(
                "/openapi/v1/namespaces",
                "POST",
                [
                    {
                        "appId": app,
                        "env": "LOCAL",
                        "clusterName": cluster,
                        "appNamespaceName": namespace,
                    }
                ],
            )
        except APIError as error:
            if "create namespace failed for" not in str(error):
                raise
            self.request(self.ns(app, namespace, cluster) + "?fillItemDetail=false")

    def cluster(self, app, cluster):
        self.request(
            f"/openapi/v1/envs/LOCAL/apps/{app}/clusters?operator=apollo",
            "POST",
            {
                "appId": app,
                "name": cluster,
                "dataChangeCreatedBy": "apollo",
                "dataChangeLastModifiedBy": "apollo",
            },
        )
        self.namespace(app, "application", cluster)

    def put(self, app, key, value, kind=0, namespace="application", cluster="default"):
        base = self.ns(app, namespace, cluster)
        body = {
            "key": key,
            "value": str(value),
            "type": kind,
            "dataChangeCreatedBy": "apollo",
            "dataChangeLastModifiedBy": "apollo",
        }
        try:
            self.request(
                base
                + "/items/"
                + urllib.parse.quote(key, safe="")
                + "?createIfNotExists=true&operator=apollo",
                "PUT",
                body,
            )
        except APIError as error:
            if error.status != 404:
                raise
            self.request(base + "/items?operator=apollo", "POST", body)

    def text(self, app, namespace, content):
        self.request(
            self.ns(app, namespace) + "/items?operator=apollo",
            "PUT",
            {
                "appId": app,
                "env": "LOCAL",
                "clusterName": "default",
                "namespaceName": namespace,
                "format": namespace.rsplit(".", 1)[-1],
                "configText": content,
                "operator": "apollo",
            },
        )

    def release(self, app, title, namespace="application", cluster="default"):
        return self.request(
            self.ns(app, namespace, cluster) + "/releases?operator=apollo",
            "POST",
            {
                "releaseTitle": title,
                "releaseComment": "apollo-evals",
                "releasedBy": "apollo",
                "isEmergencyPublish": False,
            },
        )

    def optional(self, path, *, config=False, default=None):
        try:
            return self.request(path, config=config)
        except APIError as error:
            if error.status == 404:
                return default
            raise

    def configuration(self, app, namespace="application", cluster="default"):
        data = self.optional(f"/configs/{app}/{cluster}/{namespace}", config=True, default={})
        return data.get("configurations", {})

    def snapshot_namespace(self, app, namespace="application", cluster="default"):
        base = self.ns(app, namespace, cluster)
        items = self.optional(base + "/items?page=0&size=500", default={}).get("content", [])
        return {
            "items": items,
            "release": self.optional(base + "/releases/latest"),
            "config": self.configuration(app, namespace, cluster),
        }
